impl/src/main/java/com/okta/sdk/impl/oauth2/AccessTokenRetrieverServiceImpl.java

```diff
@@ -0,0 +1,184 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.okta.sdk.impl.oauth2;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.okta.commons.http.*;
+import com.okta.commons.http.okhttp.OkHttpRequestExecutorFactory;
+import com.okta.commons.lang.Strings;
+import com.okta.sdk.impl.config.ClientConfiguration;
+import io.jsonwebtoken.Jwts;
+import org.bouncycastle.asn1.pkcs.PrivateKeyInfo;
+import org.bouncycastle.openssl.PEMKeyPair;
+import org.bouncycastle.openssl.PEMParser;
+import org.bouncycastle.openssl.jcajce.JcaPEMKeyConverter;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.*;
+import java.net.HttpURLConnection;
+import java.nio.charset.StandardCharsets;
+import java.security.InvalidKeyException;
+import java.security.KeyPair;
+import java.security.PrivateKey;
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.Date;
+import java.util.UUID;
+
+public class AccessTokenRetrieverServiceImpl implements AccessTokenRetrieverService {
+    private static final Logger log = LoggerFactory.getLogger(AccessTokenRetrieverServiceImpl.class);
+
+    private static final ObjectMapper objectMapper = new ObjectMapper();
+
+    private ClientConfiguration clientConfiguration;
+
+    public AccessTokenRetrieverServiceImpl(ClientConfiguration clientConfiguration) {
+        this.clientConfiguration = clientConfiguration;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public OAuth2AccessToken getOAuth2AccessToken() throws IOException, InvalidKeyException {
+        log.debug("Getting OAuth2 access token for client id {} from {}",
+            clientConfiguration.getClientId(), clientConfiguration.getBaseUrl());
+
+        String signedJwt = createSignedJWT(clientConfiguration);
+        String scope = String.join(" ", clientConfiguration.getScopes());
+
+        QueryString queryString = new QueryString();
+        queryString.put("grant_type", "client_credentials");
+        queryString.put("client_assertion_type", "urn:ietf:params:oauth:client-assertion-type:jwt-bearer");
+        queryString.put("client_assertion", signedJwt);
+        queryString.put("scope", scope);
+
+        HttpHeaders httpHeaders = new HttpHeaders();
+        httpHeaders.add("Accept", "application/json");
+        httpHeaders.add("Content-Type", "application/x-www-form-urlencoded");
+
+        Request accessTokenRequest = new DefaultRequest(
+            HttpMethod.POST,
+            clientConfiguration.getBaseUrl() + "/oauth2/v1/token",
+            queryString,
+            httpHeaders,
+            new ByteArrayInputStream("".getBytes("UTF-8")),
+            0);
+
+        RequestExecutor requestExecutor = new OkHttpRequestExecutorFactory().create(clientConfiguration);
```

Comment:
Don't use this directly, other clients might be using HttpClient instead of OkHttp. This has been abstracted away for you:

See:
https://github.com/okta/okta-sdk-java/blob/c5c70778ed3976adbb81bee7490024f0b6f88237/impl/src/main/java/com/okta/sdk/impl/client/BaseClient.java#L72-L75

🤔 you could probably subclass BaseClient, but that might be overkill.
Either way, reuse the RequestExecutor (or client if you go that route).  initialize it in the constructor (or similar) so this fails fast if the client implementation is missing

Reply:
Sure.. updated code.